icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -0,0 +1,251 @@
+    private final PersonNamePattern[] snFirstPatterns;
+    private final Set<String> gnFirstLocales;
+    private final Set<String> snFirstLocales;
+    private final String initialPattern;
+    private final String initialSequencePattern;
+    private final boolean capitalizeSurname;
+    private final String foreignSpaceReplacement;
+    private final boolean formatterLocaleUsesSpaces;
+    private final PersonNameFormatter.Length length;
+    private final PersonNameFormatter.Usage usage;
+    private final PersonNameFormatter.Formality formality;
+    private final Set<PersonNameFormatter.Options> options;
+
+    public PersonNameFormatterImpl(ULocale locale,
+                                   PersonNameFormatter.Length length,
+                                   PersonNameFormatter.Usage usage,
+                                   PersonNameFormatter.Formality formality,
+                                   Set<PersonNameFormatter.Options> options) {
+        // null for `options` is the same as the empty set
+        if (options == null) {
+            options = new HashSet<>();
+        }
+
+        // save off our creation parameters (these are only used if we have to create a second formatter)
+        this.length = length;
+        this.usage = usage;
+        this.formality = formality;
+        this.options = options;
+
+        // load simple property values from the resource bundle (or the options set)
+        ICUResourceBundle rb = (ICUResourceBundle)UResourceBundle.getBundleInstance(ICUData.ICU_BASE_NAME, locale);
+        this.locale = locale;
+        this.initialPattern = rb.getStringWithFallback("personNames/initialPattern/initial");
+        this.initialSequencePattern = rb.getStringWithFallback("personNames/initialPattern/initialSequence");
+        this.capitalizeSurname = options.contains(PersonNameFormatter.Options.SURNAME_ALLCAPS);
+        this.foreignSpaceReplacement = rb.getStringWithFallback("personNames/foreignSpaceReplacement");
+        this.formatterLocaleUsesSpaces = !LOCALES_THAT_DONT_USE_SPACES.contains(locale.getLanguage());
+
+        // asjust for combinations of parameters that don't make sense in practice
+        if (usage == PersonNameFormatter.Usage.MONOGRAM) {
+            // we don't support SORTING in conjunction with MONOGRAM; if the caller passes in SORTING, remove it from
+            // the options list
+            options.remove(PersonNameFormatter.Options.SORTING);
+        } else if (options.contains(PersonNameFormatter.Options.SORTING)) {
+            // we only support SORTING in conjunction with REFERRING; if the caller passes in ADDRESSING, treat it
+            // the same as REFERRING
+            usage = PersonNameFormatter.Usage.REFERRING;
+        }
+
+        // load the actual formatting patterns-- since we don't know the name order until formatting time (it can be
+        // different for different names), load patterns for both GN-first and SN-first names.  (If the user has
```

Comment:
You might want to spell out somewhere that GN refers to "given name" for _givenFirst_ context, and SN refers to "surname" for _surnameFirst_ context.

Reply:
I thought that was self-explanatory, but I ended up just changing all the instances to say "given-first" and "surname-first" like we're doing in other places.